src/sentry/ratelimits/cardinality.py

```diff
@@ -132,3 +157,160 @@ def use_quotas(
             indicates how much quota should actually be consumed.
         """
         pass
+
+
+class RedisCardinalityLimiter(CardinalityLimiter):
+    def __init__(self, cluster: str = "default", cluster_shard_factor: int = 3) -> None:
+        self.client = redis.redis_clusters.get(cluster)
+        self.cluster_shard_factor = cluster_shard_factor
+        super().__init__()
+
+    @staticmethod
+    def _get_timeseries_key(request: RequestedQuota, hash: Hash) -> str:
+        return f"cardinality-counter-{request.prefix}-{hash}"
+
+    def _get_read_sets_keys(
+        self, request: RequestedQuota, quota: Quota, timestamp: Timestamp
+    ) -> Sequence[str]:
+        oldest_time_bucket = list(quota.iter_window(timestamp))[-1]
+        return [
+            f"cardinality-sets-{request.prefix}-{shard}-{oldest_time_bucket}"
+            for shard in range(self.cluster_shard_factor)
```

Comment:
We do not need to keep all shards for an organization. 
The set is used to count the unique hashes seen in a time window.
If the number is high enough to be close to the limit, we can assume that we can count only a sample of the hashes.
So if we decide to have a sharding factor of 4, we can actually maintain only one of the shards. When we need the count of of unique time series we get the count from that set and multiply by 4.

Obviously the sharding key must guarantee that a given hash always end in the same shard.

Reply:
argh, yeah i completely forgot to implement sampling in this PR. will fix that.

Reply:
i opted for something primitive and just added two "num shards" settings. There are `0..n` shards, and the first `m` out of that list will be stored in Redis.

This has some unfortunate implications if the hash function is biased towards certain values, but that was already true.